File: client copy/src/main/java/io/github/hulang1024/chinesechess/scene/chessplay/rule/chess/ChessR.java
package io.github.hulang1024.chinesechess.scene.chessplay.rule.chess;

import io.github.hulang1024.chinesechess.scene.chessplay.rule.ChessPosition;
import io.github.hulang1024.chinesechess.scene.chessplay.rule.HostEnum;
import io.github.hulang1024.chinesechess.scene.chessplay.rule.RoundGame;

/**
 * 车
 * @author Hu Lang
 */
public class ChessR extends AbstractChess {
    public ChessR(ChessPosition pos, HostEnum host) {
        super(pos, host);
    }

    @Override
    public boolean canGoTo(ChessPosition destPos, RoundGame game) {
        int rowOffset = destPos.row - pos.row;
        int colOffset = destPos.col - pos.col;

        // 进退都可以，但必须直线走
        if (!MoveRules.isStraightLineMove(rowOffset, colOffset, MoveRules.MAX_DISTANCE)) {
            return false;
        }

        // 只要到目标位置之间没有棋子
        if (Math.abs(rowOffset) > 0) {
            int k = rowOffset > 0 ? +1 : -1;
            for (int row = pos.row + k; row != destPos.row + k; row += k) {
                if (!game.getChessboard().isEmpty(row, pos.col)) {
                    return false;
                }
            }
        } else {
            int k = colOffset > 0 ? +1 : -1;
            for (int col = pos.col + k; col != destPos.col + k; col += k) {
                if (!game.getChessboard().isEmpty(pos.row, col)) {
                    return false;
                }
            }
        }

        return true;
    }
}

File: web-client/chinese-chess/src/scene/play/rule/chess/ChessR.ts
import ChessHost from "../chess_host";
import ChessPos from "../ChessPos";
import RoundGame from "../RoundGame";
import AbstractChess from "./AbstractChess";
import { MAX_DISTANCE, isStraightLineMove, sign } from "./move_rules";

/**
 * 车
 */
export default class ChessR extends AbstractChess {
    constructor(pos: ChessPos, host: ChessHost) {
        super(pos, host);
    }

    canGoTo(destPos: ChessPos, game: RoundGame) {
        const rowOffset = destPos.row - this.pos.row;
        const colOffset = destPos.col - this.pos.col;

        // 进退都可以，但必须直线走
        if (!isStraightLineMove(rowOffset, colOffset, MAX_DISTANCE)) {
            return false;
        }

        // 只要到目标位置之间没有棋子
        if (Math.abs(rowOffset) > 0) {
            const k = sign(rowOffset);
            for (let row = this.pos.row + k; row != destPos.row; row += k) {
                if (!game.getChessboard().isEmpty(row, this.pos.col)) {
                    return false;
                }
            }
        } else {
            const k = sign(colOffset);
            for (let col = this.pos.col + k; col != destPos.col; col += k) {
                if (!game.getChessboard().isEmpty(this.pos.row, col)) {
                    return false;
                }
            }
        }

        return true;
    }
}
